File: server/public/tools/elements/widgets/GtCheckbox.ts
import {Element, Property, Listener} from "../../polymer/Annotations";
import {PolymerElement} from "../../polymer/PolymerElement";

/**
 * <gt-checkbox [checked]="foobar">
 * <gt-checkbox [radio]="foobar" value="foobar">
 */
@Element({
	selector: "gt-checkbox",
	template: "/assets/imports/widgets.html"
})
export class GtCheckbox extends PolymerElement {
	/**
	 * Disable this checkbox
	 */
	@Property({ reflect: true })
	public disabled: boolean;

	/**
	 * Current checkbox state
	 */
	@Property({ reflect: true, notify: true })
	public checked: boolean;

	/**
	 * If defined, the checkbox act as a radio button
	 * If the radio button is selected, the value attribute
	 * will be reflected to the this attribute binding.
	 */
	@Property({ reflect: true, notify: true, observer: "RadioChanged" })
	public radio: string;

	/**
	 * The value of this radio button
	 */
	@Property({ reflect: true })
	public value: string;

	/**
	 * The type of value of this radio button
	 */
	private radio_type: string;

	/**
	 * Cast the string value of this radio to the correct
	 * type for the binding
	 */
	private castValue(): any {
		switch (this.radio_type) {
			case "string": return this.value;
			case "number": return Number(this.value);
			case "boolean":
				switch (this.value) {
					case "0":
					case "false":
						return false;
					default:
						return true;
				}
			default:
				throw new Error(`Cannot cast "${this.value}" to ${this.radio_type}`);
		}
	}

	/**
	 * Click generator
	 */
	public click() {
		this.fire("click");
	}

	/**
	 * Click listener
	 */
	@Listener("click")
	private ClickEvent(e: Event) {
		if (this.disabled) {
			e.stopImmediatePropagation();
			e.preventDefault();
			return false;
		}
		else if (this.value) this.radio = this.castValue();
		else this.checked = !this.checked;
	}

	/**
	 * The radio value changed
	 * Check if we need to active the checkbox
	 */
	private RadioChanged() {
		this.radio_type = typeof this.radio;
		this.checked = this.radio == this.value;
	}
}

File: client/src/gt/component/widget/form/GtCheckbox.scala
package gt.component.widget.form

import gt.component.GtHandler
import gt.component.widget.form.GtCheckbox.AnyZero
import org.scalajs.dom.MouseEvent
import rx.{Obs, Var}
import util.Zero
import util.implicits._
import util.jsannotation.js
import xuen.Component

object GtCheckbox extends Component[GtCheckbox](
	selector = "gt-checkbox",
	templateUrl = "/assets/imports/widgets.html"
) {
	implicit val AnyZero = new Zero[Any] {
		val zero: Any = null
	}
}

@js class GtCheckbox extends GtHandler with Interactive {
	/** Disables the input */
	// TODO: make this sane again once Scala.js DOM is fixed
	this.dyn.disabled = attribute[Boolean].dyn
	@inline def _disabled: Var[Boolean] = this.dyn.disabled.asInstanceOf[Var[Boolean]]

	/** Input label */
	val label = attribute[String]

	/**
	  * If defined, the checkbox act as a radio button
	  * If the radio button is selected, this attribute
	  * will be reflected to the value model.
	  */
	val radio = property[Any] := null

	/** Current checkbox state */
	val checked = model[Boolean]

	/** Proxy to input value */
	val value = model[Any]

	override def ready(): Unit = {
		if (hasAttribute("radio")) radio := getAttribute("radio")
	}

	checked ~> Obs { if (checked && radio.! != null) value := radio.! }
	private val sync = Obs { checked := (radio.! == value.!) }
	value ~> sync
	radio ~> sync

	def mouseenter(): Unit = setAttribute("hover", "")
	def mouseleave(): Unit = removeAttribute("hover")

	override def click(): Unit = {
		if (radio.! == null) {
			checked := !checked
		} else {
			checked := true
		}
	}

	// Capture enter key presses
	listen("click", capture = true) { e: MouseEvent =>
		e.stopPropagation()
		if (!_disabled) click()
	}
}
